[Performance Dashboard Card]: Update cached data display logic

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14436
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Please have a look at the issue https://github.com/woocommerce/woocommerce-ios/issues/14436 for more details. But, essentially, this PR allows for cached data on Performance card to be displayed correctly if it is decided that it's too early to remotely fetch data again.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Have a demo site with a newly completed Order within this month, so stats can appear
2. Start app,
3. Enable Performance card, and set range to This Month.
4. Wait until data appears properly (Orders, Visitors, and Conversion), then quickly build the app again in Xcode. 
5. This will start the app again. Ensure Visitors and Conversions values are shown.

You can also contrast this by testing the same flow in `trunk` to check the issue. In step 5, the values will be hidden.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I tested this in Simulator iPhone 16 with iOS 18.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hafizrahman committed:
Update cached data display logic in nPerformance card

hafizrahman committed:
Update release notes

hafizrahman committed:
Fix controller being mentioned twice

hafizrahman committed:
First attempt at unit test for cached data display

hafizrahman commented:
@selanthiraiyan on 96620e2f8c3b456481c5d75c6959063cffbb4f0e I attempted to create a unit test for this, but it's not working yet.

This part especially 

https://github.com/woocommerce/woocommerce-ios/pull/14438/files#diff-fd13b018fba799bd2a5dbc78450971395882bf27f2aa290f21aa54d8e66a9d74R408-R409

```
        mockStorageManager.insertSampleSiteVisitStats(siteVisitStats)
        mockStorageManager.insertSampleOrderStats(orderStats)
```

Does not seem to insert the mock data correctly, while `mockStorageManager.insertSampleSiteSummaryStats(summaryStats)` is working fine.

I can see the insert is not working because when `StoreStatsPeriodViewModel.noDataFound` is being called, `siteStatsResultsController.isEmpty || orderStatsResultsController.isEmpty` are both true. Meanwhile `summaryStatsResultsController.isEmpty` returns false.

I can't figure out how to confirm whether the mock data is saved correctly or not. Because this is a unit test, I can't use an app to check the CoreData sqlite value. Any idea how to check that?

selanthiraiyan commented:
👋 @hafizrahman, 

> I can't figure out how to confirm whether the mock data is saved correctly or not. Because this is a unit test, I can't use an app to check the CoreData sqlite value. Any idea how to check that?

You could use [this method](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Storage/Storage/Extensions/NSManagedObjectContext%2BStorage.swift#L16) to load all stored objects of a particular kind and check the count. [Example](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Yosemite/YosemiteTests/Stores/ProductStoreTests.swift#L1196-L1197)

You can wrap the above check inside a `waitUntil` block to proceed after the storage is ready with saved items. What do you think? Does this help?

hafizrahman committed:
Merge branch 'trunk' into issue/14436-performance-card-display-cached-content

hafizrahman committed:
Refactor unit test.

For mock inserting data it now uses something similar to what's in StoreStatsPeriodViewModelTests.

hafizrahman commented:
@selanthiraiyan Thanks for the suggestion. While working on your suggestion, I discovered that `StoreStatsPeriodViewModelTests` has helper methods for saving, similar to what's needed in the unit test here, so I end up reusing that. https://github.com/woocommerce/woocommerce-ios/pull/14438/commits/b39cee05167c412f439734ad9d6b977d5c03b0bb

This is good for reviewing now.

selanthiraiyan reviewed (commented):
Thanks for working on this, Hafiz!

I tested by using "This Year" as the time range filter value. 

I noticed that after relaunching the app by building from Xcode visitors or conversion data is not loaded properly. 

https://github.com/user-attachments/assets/31e5ea38-3ebb-4a7c-af70-31a5df6c27fe

Please note that the data is not loaded after relaunching the app by building. I had to change the time range to get the visitors or conversion data loaded.

hafizrahman commented:
@selanthiraiyan 

> I tested by using "This Year" as the time range filter value.

This seems to be a new bug that is uncovered after the duplicate storage was fixed on 29a8deb2565f23a4edd2f98b6b69778bd74152d4

I can replicate it like so:
1. Start app, login to site freshly (can switch from another site to it),
2. Switch to This Year in Performance card,
3. Confirm that the data is shown correctly,
4. Rebuild app,
5. Check Performance card again, confirm that the visitors and conversion data are now redacted.

From debugging it seems to have to do with `endOfYear()` logic here:

https://github.com/woocommerce/woocommerce-ios/blob/09e8f0e1f2acd92d5dca596935ba821185e74904/WooFoundation/WooFoundation/Extensions/Date%2BStartAndEnd.swift#L146-L156

So when rebuilding app and using the "This Year" time range, when this is called:

<img width="1226" alt="Screenshot 2024-11-19 at 18 29 25" src="https://github.com/user-attachments/assets/d840d565-1084-4b08-b28b-dae1a20e8b5e">

`timeRange.latestDate` returns 2025-01-01, while the existing cached data is set correctly at 2024-31-12. This then resulted in nothing being found in database, so nothing is shown. I'm still investigating further.

hafizrahman committed:
Immediately show data if timestamp is fresh, since it can be assumed that data is available.

hafizrahman commented:
@selanthiraiyan I investigated the issue further and can confirm that:

- I can replicate on your test site, but not on my test site.
- the issue seems to be triggered when the timezone discrepancy is different between device timezone and site timezone (e.g: my device timezone is UTC+7 and the site timezone is UTC-6).
- On my test site when both site and device timezone is UTC+7, there is no issue.
- The problem seems to be that on initial load, the site fetches using device timezone because `ServiceLocator.stores.sessionManager.defaultSite?.siteTimezone` doesn't exist yet. Then on rebuild, the site timezone now exists and is used to get cached data, but the calculation becomes off by 1. As the timezone is used for the query to fetch site summary data, it then fails to fetch correctly.
- The issue also happens in `trunk`, meaning it is not introduced by this PR.

Given that, what do you think if with this PR you **re-review with a site that has a similar timezone as device timezone**? Because the screenshot PR depends on this PR, I think it would be better if we merge this PR first. It already contains a partial fix for sites where device and site timezone is the same, which is enough for the screenshot fix.

I will open a new issue for the timezone discrepancy problem, and it can be fixed separately. What do you think?

hafizrahman committed:
Merge branch 'trunk' into issue/14436-performance-card-display-cached-content

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/StoreStats/StorePerformanceViewModel.swift:
```diff
@@ -162,6 +162,15 @@ final class StorePerformanceViewModel: ObservableObject {
         // Stop if data is relatively new
         if !forceRefresh && DashboardTimestampStore.isTimestampFresh(for: .performance, at: timeRange.timestampRange) {
+
+            // We can show the cached data since we know it's available from the timestamp.
+            switch timeRange {
+            case .custom:
+                updateSiteVisitStatModeForCustomRange()
+            case .today, .thisWeek, .thisMonth, .thisYear:
+                siteVisitStatMode = .default
+            }
```
❓ To avoid duplicating code, could we move the logic from [these lines](https://github.com/woocommerce/woocommerce-ios/blob/c297a96f51a2274202e56905d43dbc7003d465c5/WooCommerce/Classes/ViewRelated/Dashboard/StoreStats/StorePerformanceViewModel.swift#L188-L197) into a helper method and use that?

As an added advantage I see that the original login updates the widget in `today` case.

hafizrahman replied:
I updated it in https://github.com/woocommerce/woocommerce-ios/pull/14438/commits/3a02b98c10ad668ea5c7c17e836a6d3cbf6905be

I put the Widget updating outside of the helper function, though, since it has nothing to do with `siteVisitStatMode` that the function is handling.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Dashboard/StoreStats/StorePerformanceViewModelTests.swift:
```diff
@@ -346,6 +362,66 @@ final class StorePerformanceViewModelTests: XCTestCase {
+        // Then
+        XCTAssertEqual(viewModel.siteVisitStatMode, .default) // Should be changed from `.hidden`
+
+        // Cleanup
+        DashboardTimestampStore.removeTimestamp(for: .performance, at: .today)
```
Good work cleaning this up. 

Opinion not related to PR: 

It would be nice if `DashboardTimestampStore` takes in a userDefaults instance instead of using static methods. We should be able to inject a mock user defaults and not worry about cleaning it up. Feel free to keep it as is for now.

selanthiraiyan reviewed (approved):
> Given that, what do you think if with this PR you re-review with a site that has a similar timezone as device timezone? Because the screenshot PR depends on this PR, I think it would be better if we merge this PR first. It already contains a partial fix for sites where device and site timezone is the same, which is enough for the screenshot fix.

> I will open a new issue for the timezone discrepancy problem, and it can be fixed separately. What do you think?

I agree. 

Thanks for working on this, Hafiz! Let us ship this fix to unblock screenshots work and log an issue. 🚀

hafizrahman committed:
Add helper function for updating SiteVisitStatMode
